Update substring search to use the Two Way algorithm

bluss opened the pull request:
Update substring search to use the Two Way algorithm

To improve our substring search performance, revive the two way searcher
and adapt it to the Pattern API.

Fixes #25483, a performance bug: that particular case now completes faster
in optimized rust than in ruby (but they share the same order of magnitude).

Many thanks to @gereeter who helped me understand the reverse case
better and wrote the comment explaining `next_back` in the code.

I had quickcheck to fuzz test forward and reverse searching thoroughly.

The two way searcher implements both forward and reverse search,
but not double ended search. The forward and reverse parts of the two
way searcher are completely independent.

The two way searcher algorithm has very small, constant space overhead,
requiring no dynamic allocation. Our implementation is relatively fast,
especially due to the `byteset` addition to the algorithm, which speeds
up many no-match cases.

A bad case for the two way algorithm is:

```
let haystack = (0..10_000).map(|_| "dac").collect::<String>();
let needle = (0..100).map(|_| "bac").collect::<String>());
```

For this particular case, two way is not much faster than the naive
implementation it replaces.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @aturon (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way Github handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

bluss commented:
cc @BurntSushi, you know substring search very well, cc @Kimundi about the pattern API

bluss commented:
I'll show some benchmark results. My workspace for this is temporarily [hosted as a repo](https://github.com/bluss/scratchspace/tree/twoway), if you want to check the quickcheck & benchmark implementations.

`find` and `rfind` are current libcore's `str::find(&str)` and `str::rfind(&str)`. `tw_contains` is two-way forward search, `tw_contains_rev` is two-way reverse search and `tw_paper` is an unmodified transcription of two-way from the paper.

I got a bit pessimistic when I tried out the different bad cases I could find. The performance bug #25483 is the benchmark called aa_100k below. From experience, the main improvemenet in runtime for that benchmark comes from the `byteset` optimization, not due to the TW algorithm itself.

Look at the `short_short` and `short_long` benchmarks for something of a more normal case for substring search.

```
running 45 tests
test bad_naive::find                                    ... bench:       2,553 ns/iter (+/- 112) = 28 MB/s
test bad_naive::rfind                                   ... bench:       2,179 ns/iter (+/- 35) = 33 MB/s
test bad_naive::tw_contains                             ... bench:         952 ns/iter (+/- 260) = 77 MB/s
test bad_naive::tw_contains_rev                         ... bench:         342 ns/iter (+/- 21) = 216 MB/s
test bad_naive::tw_paper                                ... bench:         587 ns/iter (+/- 5) = 126 MB/s
test bad_naive_reversed::find                           ... bench:         691 ns/iter (+/- 8) = 107 MB/s
test bad_naive_reversed::rfind                          ... bench:         690 ns/iter (+/- 133) = 107 MB/s
test bad_naive_reversed::tw_contains                    ... bench:         336 ns/iter (+/- 3) = 220 MB/s
test bad_naive_reversed::tw_contains_rev                ... bench:         735 ns/iter (+/- 21) = 100 MB/s
test bad_naive_reversed::tw_paper                       ... bench:         591 ns/iter (+/- 159) = 125 MB/s
test pathological_aa_100k::find                         ... bench:     999,765 ns/iter (+/- 117,750) = 100 MB/s
test pathological_aa_100k::rfind                        ... bench:   1,001,302 ns/iter (+/- 1,269) = 99 MB/s
test pathological_aa_100k::tw_contains                  ... bench:      16,289 ns/iter (+/- 18) = 6139 MB/s
test pathological_aa_100k::tw_contains_rev              ... bench:       7,196 ns/iter (+/- 197) = 13896 MB/s
test pathological_aa_100k::tw_paper                     ... bench:     706,485 ns/iter (+/- 655) = 141 MB/s
test pathological_aab_100k::find                        ... bench:   4,097,780 ns/iter (+/- 117,012) = 73 MB/s
test pathological_aab_100k::rfind                       ... bench:   3,814,527 ns/iter (+/- 15,899) = 78 MB/s
test pathological_aab_100k::tw_contains                 ... bench:   2,388,729 ns/iter (+/- 11,689) = 125 MB/s
test pathological_aab_100k::tw_contains_rev             ... bench:   3,354,270 ns/iter (+/- 4,309) = 89 MB/s
test pathological_aab_100k::tw_paper                    ... bench:   2,225,404 ns/iter (+/- 93,473) = 134 MB/s
test pathological_two_way_10k::find                     ... bench:     297,304 ns/iter (+/- 24,553) = 100 MB/s
test pathological_two_way_10k::rfind                    ... bench:     297,562 ns/iter (+/- 358) = 100 MB/s
test pathological_two_way_10k::tw_contains              ... bench:     300,044 ns/iter (+/- 447) = 99 MB/s
test pathological_two_way_10k::tw_contains_rev          ... bench:       3,358 ns/iter (+/- 56) = 8933 MB/s
test pathological_two_way_10k::tw_paper                 ... bench:     223,687 ns/iter (+/- 7,375) = 134 MB/s
test pathological_two_way_20k::find                     ... bench:   1,001,881 ns/iter (+/- 117,024) = 99 MB/s
test pathological_two_way_20k::rfind                    ... bench:     997,402 ns/iter (+/- 2,166) = 100 MB/s
test pathological_two_way_20k::tw_contains              ... bench:     919,003 ns/iter (+/- 2,327) = 108 MB/s
test pathological_two_way_20k::tw_contains_rev          ... bench:       5,750 ns/iter (+/- 26) = 17391 MB/s
test pathological_two_way_20k::tw_paper                 ... bench:     727,667 ns/iter (+/- 14,032) = 137 MB/s
test pathological_two_way_20k_reversed::find            ... bench:     738,045 ns/iter (+/- 772) = 81 MB/s
test pathological_two_way_20k_reversed::rfind           ... bench:     706,245 ns/iter (+/- 2,762) = 84 MB/s
test pathological_two_way_20k_reversed::tw_contains     ... bench:       5,829 ns/iter (+/- 36) = 10293 MB/s
test pathological_two_way_20k_reversed::tw_contains_rev ... bench:     295,658 ns/iter (+/- 295) = 202 MB/s
test pathological_two_way_20k_reversed::tw_paper        ... bench:     445,339 ns/iter (+/- 9,289) = 134 MB/s
test short_long::find                                   ... bench:      29,575 ns/iter (+/- 88) = 86 MB/s
test short_long::rfind                                  ... bench:      30,014 ns/iter (+/- 351) = 84 MB/s
test short_long::tw_contains                            ... bench:       7,733 ns/iter (+/- 100) = 329 MB/s
test short_long::tw_contains_rev                        ... bench:       5,910 ns/iter (+/- 41) = 431 MB/s
test short_long::tw_paper                               ... bench:      15,730 ns/iter (+/- 27) = 162 MB/s
test short_short::find                                  ... bench:         212 ns/iter (+/- 1) = 264 MB/s
test short_short::rfind                                 ... bench:         400 ns/iter (+/- 123) = 140 MB/s
test short_short::tw_contains                           ... bench:         164 ns/iter (+/- 0) = 341 MB/s
test short_short::tw_contains_rev                       ... bench:         169 ns/iter (+/- 1) = 331 MB/s
test short_short::tw_paper                              ... bench:         421 ns/iter (+/- 3) = 133 MB/s
```

gereeter commented on src/libcore/str/pattern.rs:
```diff
@@ -630,3 +509,420 @@ impl<'a, F> Pattern<'a> for F where F: FnMut(char) -> bool {
+        'search: loop {
+            // Check that we have room to search in
+            if self.position + needle.len() > haystack.len() {
+                if start == haystack.len() {
```
Should this check be done outside of the loop? We will only return `Done` on the first iteration, if ever.

bluss replied:
There is no real loop -- it's used to return back to this point. So this if check can be visited multiple times per call to next.

bluss replied:
hm I see now, that's a good point. it evolved into this.

gereeter commented on src/libcore/str/pattern.rs:
```diff
@@ -630,3 +509,420 @@ impl<'a, F> Pattern<'a> for F where F: FnMut(char) -> bool {
+
+            // See if the left part of the needle matches
+            let start = self.crit_pos;
+            for (i, &b) in needle[..start].iter().enumerate().rev() {
```
Could either this be changed to an iteration over a range or the loop in the forward search be changed to using `.iter().enumerate()`? It would be nice to keep the two sides consistent.

bluss replied:
Yes, I'll update this.

I'd like to express it with a zipped loop, but unfortunately that's much worse for performance, even with some unreachable() hints. So I'll use indexed loops here, it reads very nice.

gereeter commented on src/libcore/str/pattern.rs:
```diff
@@ -630,3 +509,420 @@ impl<'a, F> Pattern<'a> for F where F: FnMut(char) -> bool {
+    //
+    // Note: No documentation has been used for the algorithm in `next_back()`,
+    // simply trying to adapt forward search to backward search with extensive
+    // testing.
```
Quick proof that this approach should work:

All the definitions are completely symmetrical, with `period(x) = period(reverse(x))` and `local_period(u, v) = local_period(reverse(v), reverse(u))`, so if `(u, v)` is a critical factorization, so is `(reverse(v), reverse(u))`. Similarly, the "period" stored in `self.period` is the real period if `long_period` is `false`, and so is still valid for a reversed needle, and if `long_period` is true, all the algorithm requires is that `self.period` is less than or equal to the real period, which must be true for the forward case anyway, so the value stored in `self.period` is still valid.

To search in reverse through the haystack, we search forward through a reversed haystack with a reversed needle, and the above paragraph shows that the precomputed parameters can be left alone, which is exactly what this code implements.

bluss replied:
This helps! Maybe I should include this too.

It helped me realize the correct way to write the skip for the reverse case, when we check the left part (`u`).

That helped only one benchmark: The pathology when searching for `aaabaaab..` in haystack `aabaab..`.

bluss replied:
Can I steal your text whole? :-)

gereeter commented on src/libcore/str/pattern.rs:
```diff
@@ -630,3 +509,420 @@ impl<'a, F> Pattern<'a> for F where F: FnMut(char) -> bool {
+            }
+
+            // See if the left part of the needle matches
+            let start = self.crit_pos;
```
It is a conservative, and therefore completely correct, choice to assume that the period is long, as is done here. However, it might be more efficient to mirror the treatment of `memory` that is done in the forward search process.

bluss replied:
I want to leave that for later improvements

gereeter commented on src/libcore/str/pattern.rs:
```diff
@@ -630,3 +509,420 @@ impl<'a, F> Pattern<'a> for F where F: FnMut(char) -> bool {
+                byteset: byteset,
+
+                position: 0,
+                end: !0,
```
Can there be a comment explaining that this is just a dummy value that will be overwritten with the haystack length in time for searching? Also, would it be better to just use `0` or `usize::MAX`?

bluss replied:
yes I'll fix this

gereeter commented on src/libcore/str/pattern.rs:
```diff
@@ -630,3 +509,420 @@ impl<'a, F> Pattern<'a> for F where F: FnMut(char) -> bool {
+        // needle, and we want to determine whether u is a suffix of
+        // &v[..period]. If it is, we use "Algorithm CP1". Otherwise we use
+        // "Algorithm CP2", which is optimized for when the period of the needle
+        // is large.
```
I think it would be more clear to explain that when the period is large, it is expensive to compute, so `period` does not actually store the period of the needle. If `period` were the actual period, the condition would always evaluate to true by the definition of period. We simply accept a possible approximation to the actual period, but the algorithm has to be slightly adjusted (not using the "prefix memoization" with the `memory` variable) to still be correct if `period` is not completely correct.

bluss replied:
I'll add this

gereeter commented on src/libcore/str/pattern.rs:
```diff
@@ -630,3 +509,420 @@ impl<'a, F> Pattern<'a> for F where F: FnMut(char) -> bool {
+                // valid UTF-8
+                // *Rejects* however are not following character boundaries
+                let is_long = searcher.memory == usize::MAX;
+                searcher.next(self.haystack.as_bytes(), self.needle.as_bytes(), is_long)
```
Earlier, this had a branch and two calls to `next` as in
```
if is_long {
    searcher.next(/*..*/, true)
} else {
    searcher.next(/*..*/, false)
}
```
in order to encourage the creation of specialized versions of the next function, one for long periods and one for short periods. If I remember correctly, this resulted in a notable performance improvement. Have you tried seeing if this helps the new code?

bluss replied:
Aha. I tried it now and it results in a significant decrease in perforrmance actually

bluss replied:
Instead I tried just hardcoding is_long == true and never using the other case. That results in a performance gain on all the benchmarks (which may not properly cover this case though).

gereeter commented on src/libcore/str/pattern.rs:
```diff
@@ -630,3 +509,420 @@ impl<'a, F> Pattern<'a> for F where F: FnMut(char) -> bool {
+            }
+
+            // See if the right part of the needle matches
+            let start = if long_period { self.crit_pos }
```
Could the searching of the left and right parts be put into their own blocks? The shadowing of `start` seems very confusing, especially as the three instances of the same name refer to different things, and separate blocks would remove that shadowing.

bluss replied:
I renamed the first occurence of start, with that I think it should be clearer.

gereeter commented on src/libcore/str/pattern.rs:
```diff
@@ -630,3 +509,420 @@ impl<'a, F> Pattern<'a> for F where F: FnMut(char) -> bool {
+                if needle[i] != haystack[self.position + i] {
+                    self.position += self.period;
+                    if !long_period {
+                        self.memory = needle.len() - self.period;
```
It occurs to me that `self.memory` is only ever `0` or `needle.len() - self.period`, so it seems pointless to calculate `cmp::max(self.crit_pos, self.memory)` every time. Instead, `self.memory` could just be used directly, and every time it is currently set to zero it would be set to `self.crit_pos`. `needle.len() - self.period` should always be greater than `self.crit_pos`, so it needs no modification.

This is a deviation from the papers I've seen, but it should work and might result in a speedup.

bluss replied:
`let start = if long_period { 0 } else { self.memory };` would need an adjustment too in that case, and it when I simply use 0 here I get even or net loss in performance.

gereeter replied:
Ah - I missed that case. It might still work to check whether `self.memory == self.crit_pos`, and only if that is true do the loop for searching the left part, but an extra branch could easily hurt performance.

bluss commented:
Thanks for your extensive review @gereeter! I've pushed a commit to address some of this, with some cleanup. I also realized I could move the break for `Reject` until after the fast reject by the byteset. This improves the benchmarks for the forward search uniformly.

bluss commented:
The benchmark list was very long, so here's just some improvements after the later Rejects

```
commit 1:
test pathological_aa_100k::tw_contains                  ... bench:      16,289 ns/iter (+/- 18) = 6139 MB/s
test pathological_aa_100k::tw_contains_rev              ... bench:       7,196 ns/iter (+/- 197) = 13896 MB/s
test short_long::tw_contains                            ... bench:       7,733 ns/iter (+/- 100) = 329 MB/s
test short_long::tw_contains_rev                        ... bench:       5,910 ns/iter (+/- 41) = 431 MB/s
test short_short::tw_contains                           ... bench:         164 ns/iter (+/- 0) = 341 MB/s
test short_short::tw_contains_rev                       ... bench:         169 ns/iter (+/- 1) = 331 MB/s

commit 1 + 2:
test pathological_aa_100k::tw_contains                  ... bench:       3,533 ns/iter (+/- 420) = 28304 MB/s
test pathological_aa_100k::tw_contains_rev              ... bench:       3,840 ns/iter (+/- 252) = 26041 MB/s
test short_long::tw_contains                            ... bench:       3,322 ns/iter (+/- 18) = 767 MB/s
test short_long::tw_contains_rev                        ... bench:       3,343 ns/iter (+/- 44) = 763 MB/s
test short_short::tw_contains                           ... bench:         113 ns/iter (+/- 2) = 495 MB/s
test short_short::tw_contains_rev                       ... bench:         124 ns/iter (+/- 2) = 451 MB/s
```

bluss commented:
Now I feel good about the reverse algorithm, thanks @gereeter. I've split out the add-on commits to make it easier to review hopefully, so I'll squash before merge later.

gereeter commented on src/libcore/str/pattern.rs:
```diff
@@ -630,3 +509,430 @@ impl<'a, F> Pattern<'a> for F where F: FnMut(char) -> bool {
+
+            // We have found a match!
+            let match_pos = self.end - needle.len();
+            self.end -= needle.len(); // add self.period for all matches
```
To update this comment for `next_back` and make it more clear, maybe this should read "subtract self.period instead of needle.len() for overlapping matches" (similarly, the `next` case could be changed, but with add, not subtract)

bluss replied:
I'll update this

gereeter commented on src/libcore/str/pattern.rs:
```diff
@@ -630,3 +509,430 @@ impl<'a, F> Pattern<'a> for F where F: FnMut(char) -> bool {
+            // See if the left part of the needle matches
+            let start = if long_period { 0 } else { self.memory };
+            for i in (start..self.crit_pos).rev() {
+                if needle[i] != haystack[self.position + i] {
```
It might be interesting to bind to `memcmp` and see if that results in a speedup over this naive loop.

gereeter commented:
This looks really good! Short of doing word-level things, which is messy and requires `unsafe` and should probably be in a separate PR, I'm pretty much out of ideas on how to speed this up. I might add more docs and change some things stylistically, but that can be done later, and all the content here looks good. I think this should be merged, assuming the change to the Pattern API is cleared.

bluss commented:
@gereeter I found a paper on that kind of extension to this algorithm http://drops.dagstuhl.de/opus/volltexte/2011/3355/

gereeter commented:
That is the paper I had in mind when I made my comment - I saw your comment on #25483.

bluss commented:
Ok, the twoway search can be faster if instead of emitting Rejects, we just search to the next match. Let the StrSearcher handle emitting the implied intermediate Rejects.

bluss commented:
Results of moving reject handling out of the algorithm, that gave it some air to optimize the loop to find a match better. The reverse case is sped up the same way, showing results just for the forward case.

```
before:
test bad_naive::tw_contains                             ... bench:         796 ns/iter (+/- 191) = 92 MB/s
test bad_naive_reversed::tw_contains                    ... bench:         297 ns/iter (+/- 9) = 249 MB/s
test pathological_aa_100k::tw_contains                  ... bench:       3,512 ns/iter (+/- 582) = 28473 MB/s
test pathological_aab_100k::tw_contains                 ... bench:   1,951,935 ns/iter (+/- 57,089) = 153 MB/s
test pathological_two_way_10k::tw_contains              ... bench:     241,923 ns/iter (+/- 4,562) = 123 MB/s
test pathological_two_way_20k::tw_contains              ... bench:     749,007 ns/iter (+/- 39,225) = 133 MB/s
test pathological_two_way_20k_reversed::tw_contains     ... bench:       3,216 ns/iter (+/- 55) = 18656 MB/s
test short_long::tw_contains                            ... bench:       3,338 ns/iter (+/- 16) = 764 MB/s
test short_short::tw_contains                           ... bench:         112 ns/iter (+/- 1) = 499 MB/s

after:
test bad_naive::tw_contains                             ... bench:         418 ns/iter (+/- 201) = 177 MB/s
test bad_naive_reversed::tw_contains                    ... bench:         263 ns/iter (+/- 22) = 281 MB/s
test pathological_aa_100k::tw_contains                  ... bench:       3,386 ns/iter (+/- 49) = 29533 MB/s
test pathological_aab_100k::tw_contains                 ... bench:   1,004,961 ns/iter (+/- 66,086) = 298 MB/s
test pathological_two_way_10k::tw_contains              ... bench:     114,635 ns/iter (+/- 28,467) = 261 MB/s
test pathological_two_way_20k::tw_contains              ... bench:     362,096 ns/iter (+/- 34,481) = 276 MB/s
test pathological_two_way_20k_reversed::tw_contains     ... bench:       3,209 ns/iter (+/- 33) = 18697 MB/s
test short_long::tw_contains                            ... bench:       2,398 ns/iter (+/- 665) = 1063 MB/s
test short_short::tw_contains                           ... bench:         102 ns/iter (+/- 6) = 549 MB/s
```

bluss commented:
Updated to add two new commits. Will squash to one commit after review is done, and I'll update the commit logs with the fact that we don't need to change the Pattern API.

gereeter commented on src/libcore/str/pattern.rs:
```diff
@@ -630,3 +507,464 @@ impl<'a, F> Pattern<'a> for F where F: FnMut(char) -> bool {
+                // *Rejects* fall on the same indices (the intervals between matches)
+                // so they are always on character boundaries.
+                if let Some((a, b)) = last_match_fw.take() {
+                    return SearchStep::Match(a, b)
```
I believe the standard style is to have semicolons after return statements (here and in the backwards case).

Kimundi commented:
> Ok, the twoway search can be faster if instead of emitting Rejects, we just search to the next match. Let the StrSearcher handle emitting the implied intermediate Rejects. I'm going to incorporate that now, because it means we don't have to change the pattern API's specification.

@bluss: Hm, the reason @BurntSushi is emitting the `Reject` based on each `Match` is that afaik that was the only way that works with his implementation. The reason for `Reject` existing is so that consumers of the `Pattern` API can discover the boundaries of matched areas in a string without needing to search ahead to the next match or the end of the string. 

If the algorithm gets better performance by not emitting `Reject` or `Match`, then the intention is to overwrite the corresponding `Searcher` methods that don't return a `SearchStep`.

bluss commented:
@Kimundi That's a good point. What do you mean by “discover boundaries of matched areas” though, since the boundary of the match itself is clear?

Specializing the trait methods makes sense to me, but it would have been perfect if the `next_match` methods had already optimized into this, but it appears they don't. Almost identical duplicate versions of the `next` and `next_back` methods in `TwoWaySearcher` doesn't sound that appealing.

gereeter commented on src/libcore/str/pattern.rs:
```diff
@@ -630,3 +507,464 @@ impl<'a, F> Pattern<'a> for F where F: FnMut(char) -> bool {
+            // Note: add self.period instead of needle.len() to have overlapping matches
+            self.position += needle.len();
+            if !long_period {
+                self.memory = 0; // set to needle.len() - self.period for overlapping matches
```
If `self.memory` is always zero when we return a match (because we aren't looking for overlapping matches), then there isn't really any reason for it to be a member of the `TwoWaySearcher` struct.

bluss replied:
Excellent point -- I'll incorporate that depending on what we choose to do with rejects.

bluss replied:
I didn't get this to pay off, so I leave it as it is

gereeter commented:
First, that improvement is impressive enough that I think we should keep the new coalesing of `Reject`s even if it has some negative consequences.

Second, it seems that the primary users who benifit from smaller `Reject`s are `is_prefix_of` and `is_suffix_of`, which should be overridden to not use two way searching at all anyway.

bluss commented:
I just looked at benchmarks for `is_prefix_of`, and even before the latest Reject change, the results were 3 to 220 times in the old naive string search's favour.

We need to implement all the trait's method in the long run, and yes, use something else (memcmp-like) for the prefix and suffix versions.

Kimundi commented:
@bluss: Yeah, the old impls never got optimized. 
As for uses, I'm specifically referring to the `.trim_...()` methods: They skip forward until the point where the first `Reject` appears, and then stop. Example:

```
"abababcccccccccccccccccccccccccccccccccab".trim_left_matches("ab")
   => "cccccccccccccccccccccccccccccccccab"
```

If the string searcher where to be written with only `Match` instead, then the algorithm would first scan through the whole string until the final `"ab"` just to emit the start position of the `Reject`.

Having the duplication is unfortunate, but I don't see a better way apart from ensuring the generic codegen will optimize out the right cases on its own.

bluss commented:
I just addressed is_prefix_of / is_suffix_of, so they are a bit faster than previously.

Finding the *first* reject could be done simpler too, we could special case just that part maybe.

bluss commented:
For long needles, I can't make the two way searcher match the naive impl of `next_reject`, because the overhead of computing the critical factorization dominates.

Kimundi commented:
@bluss: It doesn't have to match the naive impl exactly. It can split the Rejects up at arbitrary points if it helps.

bluss commented:
I've experimented back and forth (macros, closures for specialization) and it looks like duplicating the forward and backward searches is the best unfortunately.

However, StrSearcher is a very specific algorithm for finding matches, and it is a suboptimal choice for reject-only search. In that case naive search is always better and any time we spend precomputing needle properties is wasted. For that reason, it would be best if this was given its own method on the Pattern API that circumvents the `StrSearcher`. For example `fn maximal_prefix_of(haystack: &str) -> Option<usize>` and so on.

I'm away for the weekend, so enjoy midsummer everyone!

bors commented:
:umbrella: The latest upstream changes (presumably #26192) made this pull request unmergeable. Please resolve the merge conflicts.

bluss commented:
Rebased.

Updated with:

- Use a trait to split the algorithm in two cases without duplication. Case 1 is the `MatchOnly` strategy, and the other is `RejectAndMatch`. So that we both offer the quickest possible `next_match()` impl and a general `next()` that emits `Rejects` too.
- The `RejectAndMatch` case is less streamlined for now.
- The `RejectAndMatch` case manually steps to the next character boundary so that the Pattern API can stay as it is.

bluss commented:
Here are some updated benchmarks with some caveats.

- rustc nightly is updated, llvm is updated, so things might have changed
- I can't seem to get cpu scaling to fall into the same gear again that produced the previous pretty consistent results
- So: Don't compare these literally to the previous benchmarks.

- `find:` current forward search
- `rfind:` current forward search
- `tw_contains:` new forward search in this PR
- `tw_contains_rev:` new forward search in this PR

```
running 36 tests
test bad_naive::find                                    ... bench:       1,386 ns/iter (+/- 121) = 53 MB/s
test bad_naive::rfind                                   ... bench:       1,463 ns/iter (+/- 19) = 50 MB/s
test bad_naive::tw_contains                             ... bench:         470 ns/iter (+/- 7) = 157 MB/s
test bad_naive::tw_contains_rev                         ... bench:         314 ns/iter (+/- 18) = 235 MB/s
test bad_naive_reversed::find                           ... bench:         533 ns/iter (+/- 90) = 138 MB/s
test bad_naive_reversed::rfind                          ... bench:         570 ns/iter (+/- 39) = 129 MB/s
test bad_naive_reversed::tw_contains                    ... bench:         263 ns/iter (+/- 13) = 281 MB/s
test bad_naive_reversed::tw_contains_rev                ... bench:         434 ns/iter (+/- 200) = 170 MB/s
test pathological_aa_100k::find                         ... bench:     765,670 ns/iter (+/- 51,184) = 130 MB/s
test pathological_aa_100k::rfind                        ... bench:     823,406 ns/iter (+/- 1,706) = 121 MB/s
test pathological_aa_100k::tw_contains                  ... bench:       3,893 ns/iter (+/- 32) = 25687 MB/s
test pathological_aa_100k::tw_contains_rev              ... bench:       3,848 ns/iter (+/- 97) = 25987 MB/s
test pathological_aab_100k::find                        ... bench:   2,704,123 ns/iter (+/- 99,663) = 110 MB/s
test pathological_aab_100k::rfind                       ... bench:   2,824,903 ns/iter (+/- 163,534) = 105 MB/s
test pathological_aab_100k::tw_contains                 ... bench:   1,149,895 ns/iter (+/- 73,320) = 260 MB/s
test pathological_aab_100k::tw_contains_rev             ... bench:   1,120,750 ns/iter (+/- 136,251) = 267 MB/s
test pathological_two_way_10k::find                     ... bench:     227,447 ns/iter (+/- 157) = 131 MB/s
test pathological_two_way_10k::rfind                    ... bench:     244,858 ns/iter (+/- 14,582) = 122 MB/s
test pathological_two_way_10k::tw_contains              ... bench:     136,919 ns/iter (+/- 13,430) = 219 MB/s
test pathological_two_way_10k::tw_contains_rev          ... bench:       3,208 ns/iter (+/- 18) = 9351 MB/s
test pathological_two_way_20k::find                     ... bench:     762,829 ns/iter (+/- 1,937) = 131 MB/s
test pathological_two_way_20k::rfind                    ... bench:     820,271 ns/iter (+/- 780) = 121 MB/s
test pathological_two_way_20k::tw_contains              ... bench:     426,550 ns/iter (+/- 35,134) = 234 MB/s
test pathological_two_way_20k::tw_contains_rev          ... bench:       5,263 ns/iter (+/- 10) = 19000 MB/s
test pathological_two_way_20k_reversed::find            ... bench:     505,162 ns/iter (+/- 44,729) = 118 MB/s
test pathological_two_way_20k_reversed::rfind           ... bench:     538,927 ns/iter (+/- 30,293) = 111 MB/s
test pathological_two_way_20k_reversed::tw_contains     ... bench:       3,564 ns/iter (+/- 16) = 16834 MB/s
test pathological_two_way_20k_reversed::tw_contains_rev ... bench:     179,662 ns/iter (+/- 3,419) = 333 MB/s
test short_long::find                                   ... bench:      23,059 ns/iter (+/- 118) = 110 MB/s
test short_long::rfind                                  ... bench:      24,691 ns/iter (+/- 75) = 103 MB/s
test short_long::tw_contains                            ... bench:       2,126 ns/iter (+/- 4) = 1199 MB/s
test short_long::tw_contains_rev                        ... bench:       2,040 ns/iter (+/- 16) = 1250 MB/s
test short_short::find                                  ... bench:         160 ns/iter (+/- 1) = 350 MB/s
test short_short::rfind                                 ... bench:         311 ns/iter (+/- 1) = 180 MB/s
test short_short::tw_contains                           ... bench:          81 ns/iter (+/- 58) = 691 MB/s
test short_short::tw_contains_rev                       ... bench:          99 ns/iter (+/- 0) = 565 MB/s
```

bluss commented:
I'll tentatively nominate this for the beta because I had hoped I would get this in before we branched for 1.2. It's a big performance improvement, no other perks.

gereeter commented:
I just realized that I never actually said this, so LGTM.

bluss commented:
@gereeter Thank you, you've provided a lot of valuable help & feedback!

aturon commented:
@bors: r+

Thanks for doing this!

bors commented:
:pushpin: Commit 274bb24 has been approved by `aturon`

<!-- @bors r=aturon 274bb24efdbeed0ab1a91f3c02f86551ef16eac7 -->

bors commented:
:hourglass: Testing commit 274bb24 with merge 7fc0675...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/5549), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/5561), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/5508), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/5536), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/5508), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/5582), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/5572), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/5540), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/524), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/525), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/521), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/523)

alexcrichton commented:
triage: not accepting for a beta backport
